File: app/src/main/java/com/abishov/hexocat/commons/utils/CrashReportingTree.java
package com.abishov.hexocat.commons.utils;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;

import java.util.Locale;

import hu.supercluster.paperwork.Paperwork;
import timber.log.Timber;

public final class CrashReportingTree extends Timber.Tree {

    @NonNull
    private final Paperwork paperwork;

    public CrashReportingTree(@NonNull Paperwork paperwork) {
        this.paperwork = paperwork;
    }

    @Override
    protected void log(int priority, String tag,
            @NonNull String message, @Nullable Throwable throwable) {
        if (priority == Log.VERBOSE || priority == Log.DEBUG) {
            return;
        }

        if (priority == Log.ERROR && throwable != null) {
            // log error using crash reporting tool
            Timber.tag(String.format(Locale.US, "Tag=[%s], sha=[%s], date=[%s]", tag,
                    paperwork.get("gitSha"), paperwork.get("buildDate")));
            Timber.e(throwable, message);
        }
    }
}

File: app/src/main/java/com/abishov/hexocat/common/utils/CrashReportingTree.kt
package com.abishov.hexocat.common.utils

import android.util.Log
import hu.supercluster.paperwork.Paperwork
import timber.log.Timber
import java.util.*

class CrashReportingTree(private val paperwork: Paperwork) : Timber.Tree() {

  override fun log(priority: Int, tag: String, message: String, throwable: Throwable?) {
    if (priority == Log.VERBOSE || priority == Log.DEBUG) {
      return
    }

    if (priority == Log.ERROR && throwable != null) {
      Timber.tag(
        String.format(
          Locale.US, "Tag=[%s], sha=[%s], date=[%s]", tag,
          paperwork.get("gitSha"), paperwork.get("buildDate")
        )
      )

      Timber.e(throwable, message)
    }
  }
}
